doc/mkdocs/docs/api/json_pointer/string_t.md

```diff
@@ -0,0 +1,16 @@
+# <small>nlohmann::json_pointer::</small>string_t
+```cpp
+using string_t = RefStringType;
+```
+
+The string type used for the reference tokens making up the JSON pointer.
+
+## Notes
+
+When used for array access an equivalent function to `std::stoull` compatible with `string_t` must be findable by argument-dependent lookup (ADL).
```

Comment:
> > When used for array access
> 
> Is this only required if array access is used, only if certain member functions are used, or is this always a requirement? If the last, perhaps "In order to support array access", or just leave off this conditional.

I'm trying to come up with concise wording and am struggling. As we're iterating the reference tokens of the JSON pointer and decending into the JSON object, if the current value is an array, `stoull` is called (to attempt) to convert the reference token to an array index.

> > an equivalent function to `std::stoull` compatible with `string_t` must be findable by argument-dependent lookup (ADL).
> 
> I'm stumbling over this phrasing. Maybe something like this?
> 
> "there must be a function named `stoull` that is findable by argument-dependent lookup (ADL) when called with a `string_t` first parameter, and that provides equivalent functionality to `std::stoull`."
> 
> Advanced options (sorry for the length):
> 
> Should we provide a bit more information for those not well-versed in ADL, and how the call usage affects how the matching is done? (We run into that a lot with the `to_json/from_json` functions.)

I'd consider this an advanced use case. Maybe add a general chapter on ADL and link to that? (As you say, it's useful in other places.) I'd consider that for a separate PR.

> There are two optional parameters. Is it called with just the first argument? Are one or both of the optional parameters used? Is a specific type needed for the return value or is it just something convertible to `unsigned long long`?
> 
> It looks like it is called with a variable of type `const string_t &` as the first parameter, a variable of type `std::size_t *` as the second parameter, no third parameter, and assigns the return value to an existing variable of type `unsigned long long`.
> 
> This could help the user find/create an equivalent function for this, rather than having to implement a full replacement. For example, they don't need `base` support, and as long as the return type is implicitly convertible to `unsigned long long` it will be usable. The existing variable is important in that it's not doing `auto ret = stoull()` where the return type must match exactly or it could cause compile problems later. Then again, we may want it to be 100% signature compatible in order to not break things if we change the usage in the future, and it would be good to call that out even if the compiler doesn't require it now.

Yes, I can clarify those requirements as you suggest. `base` seems unneeded unless the JSON pointer RFC is amended. Currently leading zeros are explicitly disallowed (by the code anyway, I assume that conforms to the RFC).

> Brainstorming on possible expanded wording for the current compiler-enforced requirements: There must be a `stoull` function findable by argument-dependent lookup (ADL) (located in the `std` namespace or the namespace of the type used as `RefStringType`) that provides equivalent behavior to `std::stoull`. A `const string_t &` is passed as the first parameter of this function, a `std::size_t *` is passed as the second parameter, and the optional third parameter is left as the default. The return value must be implicitly convertible to `unsigned long long`.

I'll think about it some more and come back.

Reply:
I had another idea.

Currently, we require all strings to be a sequence of 8 bit chars and have the member functions `data()` and `size()`. That would enable the following:
```cpp
unsigned long long val; // could be basic_json::size_type
auto result = std::from_chars(str.data(), str.data() + str.size(), val);
```
Limited to C++17 and above but has the following advantages:
* No customization required.
* Fast.
* Does not throw. (Long term, I'd like exceptions to be optional and errors alternatively be reported through `std::expected`.)
* Doesn't allow "/foo/     0" (leading whitespace) which is perfectly valid using `std::stoull`.

We'd still have to support `std::stoull` for C++11/C++14 and ADL but many users may not need to care.

Disadvantages:
* Different behavior based on used standard can be confusing to users.
* Two different code paths to maintain.
* Further solidifies dependence on 8 bit char strings.

Reply:
Just a naive thought: we could also add a stripped-down version of Google Abseil's `absl::StrCat` which takes care of a lot of edge cases.

Reply:
I'm sure it's a much more polished implementation than mine, but concatenation isn't the problem here?

I've switched the code from `std::stoull` to using `std::stroull` (which `std::stoull` uses internally anyway). That removes the need for an ADL findable `stoull`. Everything just works as long as the custom string type has `c_str()` (sadly `std::stroull` requires a null-terminated string).

Reply:
It can also concat numbers and strings to a string, that’s why I thought of it.

Reply:
Oh sorry, we need it the other way around. Stupid me. Sorry…

Reply:
It would still be useful to replace instances of `concat(..., std::to_string(integer), ...)` but the existing templates could be easily modified to do that if someone wanted to. I opted not to do everything at once.